### PR TITLE
[YAML] Add PubSub reading and writing transforms.

### PR DESCRIPTION
**Pull request by robertwb**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

File: sdks/python/apache_beam/transforms/core.py [hunk 1]
```diff
@@ -2272,17 +2276,24 @@ def __or__(self, transform):
...
-    if result[self.main_output_tag()].element_type == typehints.Any:
-      result[self.main_output_tag()].element_type = transform.infer_output_type(
-          self._pcoll.element_type)
-    # TODO(BEAM-18957): Add support for tagged type hints.
-    result[self.error_output_tag()].element_type = typehints.Any
-    return _PValueWithErrors(
-        result[self.main_output_tag()],
-        self._exception_handling_args,
-        self._upstream_errors + (result[self.error_output_tag()], ))
+    if hasattr(transform, 'with_exception_handling'):
```
Comment (liferoad) [inline sdks/python/apache_beam/transforms/core.py:2279]: any reason to add this if now?

File: sdks/python/apache_beam/typehints/schemas.py [hunk 2]
```diff
@@ -226,6 +226,15 @@ def option_from_runner_api(
       schema_registry=schema_registry).option_from_runner_api(option_proto)
 
 
+def schema_field(
+    name: str, field_type: Union[schema_pb2.FieldType,
```
Comment (liferoad) [inline sdks/python/apache_beam/typehints/schemas.py:230]: do we have enough tests for core.py and schemas.py?

Comment (robertwb) [reply sdks/python/apache_beam/typehints/schemas.py:230]: They should be quite well covered.

File: sdks/python/apache_beam/transforms/core.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/transforms/core.py:2279]: Yes, to support this generic capability for the Read/Write operations (that currently don't have a with_exception_handling option).

File: sdks/python/apache_beam/typehints/schemas.py (hunk 2 above)
Comment (liferoad) [reply sdks/python/apache_beam/typehints/schemas.py:230]: I am fine if the existing tests cover the new code changes well.

File: sdks/python/apache_beam/yaml/standard_io.yaml [hunk 3]
```diff
@@ -53,6 +53,8 @@
     # 'WriteToBigQuery': 'apache_beam.yaml.yaml_io.write_to_bigquery'
     'ReadFromText': 'apache_beam.yaml.yaml_io.read_from_text'
     'WriteToText': 'apache_beam.yaml.yaml_io.write_to_text'
+    'ReadFromPubSub': 'apache_beam.yaml.yaml_io.read_from_pubsub'
```
Comment (liferoad) [inline sdks/python/apache_beam/yaml/standard_io.yaml:56]: `apache_beam.yaml.yaml_io` is a little redundant. was `apache_beam.yaml.io` considered before?

File: sdks/python/apache_beam/yaml/yaml_io.py [hunk 4]
```diff
@@ -123,6 +131,134 @@ def raise_exception(failed_row_with_error):
...
+def read_from_pubsub(
+    root,
+    *,
+    topic: Optional[str] = None,
+    subscription: Optional[str] = None,
+    format: str,
+    schema: Optional[Any] = None,
+    attributes: Optional[Iterable[str]] = None,
+    attributes_map: Optional[str] = None,
+    timestamp_attribute: Optional[str] = None):
```
Comment (liferoad) [inline sdks/python/apache_beam/yaml/yaml_io.py:168]: Do we need docstrings?

File: sdks/python/apache_beam/yaml/yaml_io.py [hunk 5]
```diff
@@ -123,6 +131,134 @@ def raise_exception(failed_row_with_error):
...
+        for attr in attributes:
+          values[attr] = msg.attributes[attr]
+      if attributes_map:
+        values[attributes_map] = msg.attributes
+      return beam.Row(**values)
+
+  output = (
+      root
+      | beam.io.ReadFromPubSub(
+          topic=topic,
```
Comment (liferoad) [inline sdks/python/apache_beam/yaml/yaml_io.py:200]: any reason to ignore `id_label`?

File: sdks/python/apache_beam/yaml/yaml_io.py [hunk 6]
```diff
@@ -123,6 +131,134 @@ def raise_exception(failed_row_with_error):
...
+          subscription=subscription,
+          with_attributes=bool(attributes or attributes_map),
+          timestamp_attribute=timestamp_attribute)
+      | 'ParseMessage' >> beam.Map(mapper))
```
Comment (liferoad) [inline sdks/python/apache_beam/yaml/yaml_io.py:204]: I do not know the details here. Just question: Could ParseMessage cause any label conflicts if two pub/sub IOs are created?

File: sdks/python/apache_beam/yaml/yaml_io.py [hunk 7]
```diff
@@ -123,6 +131,134 @@ def raise_exception(failed_row_with_error):
...
+          field for field in input_schema.fields
+          if field.name not in extra_fields
+      ])
+  formatter = _create_formatter(format, schema, payload_schema)
+  return (
+      pcoll | beam.Map(
+          lambda row: beam.io.gcp.pubsub.PubsubMessage(
+              formatter(row), attributes_extractor(row)))
+      | beam.io.WriteToPubSub(
+          topic, with_attributes=True, timestamp_attribute=timestamp_attribute))
```
Comment (liferoad) [inline sdks/python/apache_beam/yaml/yaml_io.py:259]: similarly, id_label?

File: sdks/python/apache_beam/yaml/yaml_io_test.py [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+import apache_beam as beam
+from apache_beam.io.gcp.pubsub import PubsubMessage
+from apache_beam.testing.util import assert_that
+from apache_beam.testing.util import AssertThat
+from apache_beam.testing.util import equal_to
+from apache_beam.yaml import yaml_provider
+from apache_beam.yaml.yaml_transform import YamlTransform
+
+
+class FakeReadFromPubSub:
```
Comment (liferoad) [inline sdks/python/apache_beam/yaml/yaml_io_test.py:33]: Besides mock, will we do the integration tests somewhere?

File: sdks/python/apache_beam/yaml/yaml_io.py (hunk 6 above)
Comment (liferoad) [reply sdks/python/apache_beam/yaml/yaml_io.py:252]: Thanks. Got it now.

File: sdks/python/apache_beam/yaml/yaml_io_test.py [hunk 9]
```diff
@@ -50,14 +63,17 @@ def __call__(
...
-  def __init__(self, topic, messages, timestamp_attribute=None):
+  def __init__(
+      self, topic, messages, id_attribute=None, timestamp_attribute=None):
     self._topic = topic
     self._messages = messages
+    self._id_attribute = id_attribute
     self._timestamp_attribute = timestamp_attribute
 
-  def __call__(self, topic, *, with_attributes, timestamp_attribute):
+  def __call__(self, topic, *, with_attributes, id_label, timestamp_attribute):
```
Comment (liferoad) [inline sdks/python/apache_beam/yaml/yaml_io_test.py:72]: id_attribute?

Comment (liferoad) [reply sdks/python/apache_beam/yaml/yaml_io_test.py:72]: nvm. this is for the fake mock.

File: sdks/python/apache_beam/yaml/yaml_io_test.py (hunk 8 above)
Comment (liferoad) [reply sdks/python/apache_beam/yaml/yaml_io_test.py:31]: Sounds good as long as we have the plan to do it.

**Review by liferoad — APPROVED** [review]

**Comment by liferoad** [comment]
LGTM. Please fix the failed tests before merging.

**Pull request merged**